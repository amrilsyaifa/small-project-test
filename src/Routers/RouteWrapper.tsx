import React from 'react';
import { useLocation } from 'wouter';
import { IPageRouter } from './PageRouter';

interface IRouteWrapper extends IPageRouter {
  Render: React.FC<IPageRouter>;
}

/**
 * Wrap Router file so it won't load function inside the router unless it's part of the url children (startsWith)
 * Please note it's using relative location from useLocation (not the absolute path because it's discarded by Page Router)
 */
export const RouteWrapper: React.FC<IRouteWrapper> = ({ Render, base }) => {
  const [location] = useLocation();

  if (!location.startsWith(base)) return null;
  return <Render base={base} key={base} />;
};

export default RouteWrapper;
