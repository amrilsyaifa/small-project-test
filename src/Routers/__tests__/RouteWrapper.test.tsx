import { render } from '@testing-library/react';
import React from 'react';
import { useLocation } from 'wouter';
import { RouteWrapper } from '../RouteWrapper';

jest.mock('wouter', () => ({
  ...jest.requireActual('wouter'),
  useLocation: jest.fn()
}));

const mockUseLocation = useLocation as jest.Mock;

describe('RouteWrapper', () => {
  it('should render children on the right url', () => {
    mockUseLocation.mockReturnValue(['/correct']);
    const { container, queryByText } = render(
      <RouteWrapper base="/correct" Render={() => <div>This should display</div>} />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('This should display')).toBeDefined();
  });

  it('should render children on the children url', () => {
    mockUseLocation.mockReturnValue(['/correct/children']);
    const { container, queryByText } = render(
      <RouteWrapper base="/correct" Render={() => <div>This should display</div>} />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('This should display')).toBeDefined();
  });

  it('should render NOT render on the wrong url', () => {
    mockUseLocation.mockReturnValue(['/correct']);
    const { container, queryByText } = render(
      <RouteWrapper base="/wrong" Render={() => <div>This should NOT display</div>} />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('This should NOT display')).toBeNull();
  });

  it('should render NOT render on the wrong url parent', () => {
    mockUseLocation.mockReturnValue(['/wrong/correct']);
    const { container, queryByText } = render(
      <RouteWrapper base="/correct" Render={() => <div>This should NOT display</div>} />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText('This should NOT display')).toBeNull();
  });
});
