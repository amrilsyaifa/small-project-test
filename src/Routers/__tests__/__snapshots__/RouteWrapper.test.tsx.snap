// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteWrapper should render NOT render on the wrong url 1`] = `<div />`;

exports[`RouteWrapper should render NOT render on the wrong url parent 1`] = `<div />`;

exports[`RouteWrapper should render children on the children url 1`] = `
<div>
  <div>
    This should display
  </div>
</div>
`;

exports[`RouteWrapper should render children on the right url 1`] = `
<div>
  <div>
    This should display
  </div>
</div>
`;
